OS Detection and setting the old or new settings UI page accordingly

alekhyareddy28 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This PR detects if the OS version is before 19h1 or after. 
If it was an earlier build, it accordingly deploys the old settings UI page and renders the description of Launcher and keyboard remapper as follows. If not the user is redirected to the new settings UI.
Image resizer shows it's old 
![image](https://user-images.githubusercontent.com/28739210/81028465-627da100-8e36-11ea-86f3-b71d4ec24df6.png)
![image](https://user-images.githubusercontent.com/28739210/81028469-6ad5dc00-8e36-11ea-964a-4e4ab4bf68a8.png)

<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #2443 
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
* The code obtains the `build number` from the registry and checks if it is before or after the 19h1 release. It accordingly deploys the new or old settings UI executable.

## Outstanding Tasks
* Need to evaluate the differences in Image resizer in v0.17 v/s settings v2 and implement it accordingly. Shall do it as a part of a separate PR. As of now Image resizer settings show up as follows in the old settings - 
![image](https://user-images.githubusercontent.com/28739210/81107710-7247c480-8ecc-11ea-99f6-8cf1e68a5df1.png)
![image](https://user-images.githubusercontent.com/28739210/81107781-8c81a280-8ecc-11ea-9988-86fe323fd5a1.png)


<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
* Created vhd images for RS4 and RS5 releases and validated the MSI by running it on those builds.
* Link to MSI - (Internal to microsoft) https://microsoft.sharepoint.com/:u:/t/Notes145/Ebf-9tk39U1Cha7ECAGyOxQBjJiA3LSQiMRG94ng878n2Q?e=50bBIt

crutkas commented:
We should be feature detecting with API contracts.  Not OS version. The code sample I sent out did just that. It was a link into the WinUI repo

alekhyareddy28 commented:
@crutkas, I did not know that. @laviusmotileng-ms forwarded an email to me which had a discussion but it did not have any reference to the code. If you could share it with me, I'll implement it accordingly tomorrow.

alekhyareddy28 commented:
Also, @crutkas, could you clarify regarding the old settings link for Image Resizer?  

> If RS4/5
old settings
KBM disabled even if JSON says it is enabled
Launcher disabled even if JSON says it is enabled
Tab for -> Launcher: “This feature requires Windows 10, May 2019 Update”
Tab for -> KBM: “This feature requires Windows 10, May 2019 Update”
OS detect for image resizer so its old settings link still works.
This I don’t know if it will work.

arjunbalgovind commented:
> Also, @crutkas, could you clarify regarding the old settings link for Image Resizer?
> 
> > If RS4/5
> > old settings
> > KBM disabled even if JSON says it is enabled
> > Launcher disabled even if JSON says it is enabled
> > Tab for -> Launcher: “This feature requires Windows 10, May 2019 Update”
> > Tab for -> KBM: “This feature requires Windows 10, May 2019 Update”
> > OS detect for image resizer so its old settings link still works.
> > This I don’t know if it will work.

In .17 the Image Resizer settings were within the app itself, whereas for SettingsV2, the "Settings" button within Image Resizer was removed and the settings were moved to SettingsV2. I think we would have to revert the part where the Settings button was removed from the UI in #2324 and instead show that button depending on the OS. Keep in mind that Image Resizer can run even if the runner isn't running so that logic would have to be independent.

enricogior commented on src/runner/settings_window.cpp:
```diff
@@ -206,6 +206,37 @@ BOOL run_settings_non_elevated(LPCWSTR executable_path, LPWSTR executable_args,
+const int RS5_VERSION = 17999;
+
+// This function obtains the os build number from the registry and accordingly sets the settings executable
+void os_detection()
```
Instead of setting a global variable, make `os_detection` return a value. And rename it so it does auto explain what it returns.

alekhyareddy28 commented:
@crutkas and @enricogior, I updated the PR accordingly. Thanks.

riverar commented:
⚠ This is not the right way to do this. Recommend this be revisited and use the appropriate [Version Helper APIs](https://docs.microsoft.com/en-us/windows/win32/sysinfo/version-helper-apis).

riverar commented on src/runner/settings_window.cpp:
```diff
@@ -206,6 +206,37 @@ BOOL run_settings_non_elevated(LPCWSTR executable_path, LPWSTR executable_args,
+            }
+        }
+    }
+}
```
⚠ Registry handle gets leaked. RegCloseKey must be called.

riverar commented on src/runner/settings_window.cpp:
```diff
@@ -206,6 +206,37 @@ BOOL run_settings_non_elevated(LPCWSTR executable_path, LPWSTR executable_args,
+    
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE, osVersionPath, 0, KEY_READ, &hkey) == ERROR_SUCCESS)
+    {
+        wchar_t buildNumberString[256];
```
⚠ 256 is not a safe assumption here. Instead, call RegQueryValueEx twice, once with a nullptr buffer, trap on ERROR_MORE_DATA, then build an appropriate buffer and call again. (Docs have an example of [this common pattern](https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regqueryvalueexw#remarks).)

riverar commented on src/runner/settings_window.cpp:
```diff
@@ -206,6 +206,37 @@ BOOL run_settings_non_elevated(LPCWSTR executable_path, LPWSTR executable_args,
+        DWORD bufferSize = sizeof(buildNumberString);
+        int buildNum = 0;
+
+        if (RegQueryValueExW(hkey, L"CurrentBuildNumber", 0, NULL, (LPBYTE)buildNumberString, &bufferSize) == ERROR_SUCCESS)
```
Nit: Replace `NULL` with `nullptr`

riverar commented on src/runner/settings_window.cpp:
```diff
@@ -206,6 +206,37 @@ BOOL run_settings_non_elevated(LPCWSTR executable_path, LPWSTR executable_args,
 
+// This variable is set based on the os version
+bool use_old_settings = true;
+const int RS5_VERSION = 17999;
```
⚠ Wrong version. It should be `17763`.

alekhyareddy28 commented:
@riverar , I've modified the code and no longer using the Registry. I've adopted the same approach as that used by microsoft-ui-xaml. Added a link to their code base in the comments before the function.

riverar commented:
@alekhyareddy28 Well that was fast! 👍
